### ISSUE TITLE
jdbcwriter didn't work

### ISSUE THREAD

**Issue by ghost** [tinylog-org/tinylog#30]
Hello,

I try to setup a jdbcwriter but I got following errors:

LOGGER ERROR: Failed to activate configuration (java.sql.SQLException: No suitable driver found for jdbc:mysql://10.211.55.5:3306/XXXXXX)

However the mysql driver is in project path (class path) as well as in server lib (Tomcat). My application can access the mysql database.

Any suggestion? Did I miss a configuration?

Thx

**Comment by pmwmedia** [comment]
Can you try to add the line `Class.forName("com.mysql.jdbc.Driver").newInstance();` straight before configuring tinylog via the configurator class. Please exclude temporarily tinylog.properties for this test if existing.

For example:

```
Class.forName("com.mysql.jdbc.Driver").newInstance();
Configurator.defaultConfig()
   .writer(new JdbcWriter("jdbc:mysql://10.211.55.5:3306/DATABASE", "TABLE", Arrays.asList(Value.MESSAGE)))
   .activate();
```

Please let me know whether this works?

**Comment by ghost** [comment]
I tried as you recommended. I got following error:

Loading class `com.mysql.jdbc.Driver'. This is deprecated. The new driver class is`com.mysql.cj.jdbc.Driver'. The driver is automatically registered via the SPI and manual loading of the driver class is generally unnecessary.

I  skipped the Class.forName("com.mysql.jdbc.Driver").newInstance(); but the same error still occur. So I think you internally called this driver. My environment is JDBC4 and Java 1.8, mysql connector is 5.1.38. My feeling is that tinylog can’t run in this environment with jdbcwriter, right?

Also, how can i set the user and password for the connection. I tried to attach &user=XXX&password=YYY to the jdbc url. 

Kind Regards

hjlabrador

Absender: Martin Winandy notifications@github.com
Antworten: pmwmedia/tinylog reply@reply.github.com
Datum: 29. April 2016 at 17:18:57
Empfänger: pmwmedia/tinylog tinylog@noreply.github.com
CC: hjlabrador labrador@posteo.de, Author author@noreply.github.com
Betreff:  Re: [pmwmedia/tinylog] jdbcwriter didn't work (#30)  

Can you try to add the line Class.forName("com.mysql.jdbc.Driver").newInstance(); straight before configuring tinylog via the configurator class. Please exclude temporarily tinylog.properties for this test if existing.

For example:

Class.forName("com.mysql.jdbc.Driver").newInstance();
Configurator.defaultConfig()
   .writer(new JdbcWriter("jdbc:mysql://10.211.55.5:3306/DATABASE", "TABLE", Arrays.asList(Value.MESSAGE)))
   .activate();

Please let me know whether this works?

**Comment by pmwmedia** [comment]
JdbcWriter just calls `DriverManager.getConnection(url)` (see: [JdbcWriter.java:319](https://github.com/pmwmedia/tinylog/blob/v1.1/tinylog/src/org/pmw/tinylog/writers/JdbcWriter.java#L319)). This is standard Java API and should work also on your environment. Actually JdbcWriter is used quite often together with MySQL. What happens if you call `DriverManager.getConnection(url)` directly?

You can pass username and password via arguments: [Javadoc JdbcWriter](http://www.tinylog.org/javadoc/org/pmw/tinylog/writers/JdbcWriter.html). But `?user=X&password=Y` should work as well.

Can you share a small example project for reproducing the issue? Without example code I can just guess.

**Comment by ghost** [comment]
So, I could get it to work with Class.forName("com.mysql.jdbc.Driver").newInstance(); ahead the Configurator as you recommended. However with tinylog.properties I still have the problem that the driver could not be found. Maybe you have an idea about your implementation. So far I can go on with my project, however the properties is a better way for configuration.

HJLabrador

Kind Regards

Absender: Martin Winandy notifications@github.com
Antworten: pmwmedia/tinylog reply@reply.github.com
Datum: 30. April 2016 at 11:58:02
Empfänger: pmwmedia/tinylog tinylog@noreply.github.com
CC: hjlabrador labrador@posteo.de, Author author@noreply.github.com
Betreff:  Re: [pmwmedia/tinylog] jdbcwriter didn't work (#30)  

JdbcWriter just calls DriverManager.getConnection(url) (see: JdbcWriter.java:319). This is standard Java API and should work also on your environment. Actually JdbcWriter is used quite often together with MySQL. What happens if you call DriverManager.getConnection(url) directly?

You can pass username and password via arguments: Javadoc JdbcWriter. But ?user=X&password=Y should work as well.

Can you share a small example project for reproducing the issue? Without example code I can just guess.

**Comment by pmwmedia** [comment]
I have set up a new tomcat installation from scratch. Now I'm able to reproduce the problem. Out of the box, tomcat can't load driver services. You have either to create an instance of the driver or to declare a data source. I plan to support data sources in tinylog 1.2. Help is welcome for the implementation :-)

For tinylog 1.1, I propose to load the driver via a `WebListener`:

```
@WebListener
public class LifeCycleListener implements ServletContextListener {

    @Override
    public void contextInitialized(ServletContextEvent event) {
        try {
            new com.mysql.jdbc.Driver();
        } catch (SQLException e) {
            Logger.error(e);
        }
    }

    @Override
    public void contextDestroyed(ServletContextEvent event) {
    }

}
```

The configuration can be loaded from `tinylog.properties`. You don't need to use the configurator class. It is only imported that the JDBC driver is loaded before any logger call.

**Comment by pmwmedia** [comment]
Fixed in tinylog 1.2 branch by supporting DataSources
